### ISSUE TITLE
why not write transducers?

### ISSUE THREAD

**Issue by brucou** [tomi/fromfrom#30]
The synchronous transformation of data has already been solved elegantly with transducers. You can refer to:

- https://jlongster.com/Transducers.js--A-JavaScript-Library-for-Transformation-of-Data
- https://github.com/thi-ng/umbrella/tree/master/packages/transducers (a more complete transducers library)

Transducers apply equally to iterators and observable. I don't mean this to belittle your effort, I rather want to attract the attention on prior art and not reinventing (redebugging) the wheel.

One possible advantage of your approach may be the typings. Transducers are pretty hard to type both accurately and concisely.

**Comment by x2764tech** [comment]
When I saw this library, I immediately thought of [lodash's chain](https://lodash.com/docs#chain) function:

    _.chain([1,2,3]).filter(x => x % 2).value

**Comment by taverasmisael** [comment]
This is also a good article (Book chapter?) by Eric Elliot on the topic https://medium.com/javascript-scene/transducers-efficient-data-processing-pipelines-in-javascript-7985330fe73d

**Comment by tomi** [comment]
Hi and thank you for your comments.

The question about transducers is valid and it's something I could have commented on for example in the readme. With this library I wanted to have a library with a clean and simple API and a great developer experience with good typings. Hence I opted with the single export and a fluent API. Now I am well aware of the downsides of this approach. Mainly that it doesn't play well with tree shaking and it's not that easy to extend. Now while transducers do solve those two issues, they are, like you described, harder to type and don't always provide the best developer experience. If we look at the example @taverasmisael provided in issue #28 :

```ts
// from the main example
import { compose, filter, sortBy } from 'fromfrom/transforms';

const activeSortedByName = compose(
  sortByDescending(user => user.age),
  filter(user => user.active)
);

const result = activeSortedByName(users).toArray();
```

When you type that `sortByDescending(` in your editor, it has no knowledge what type of the `user` parameter should be (unless you label it yourself) and you don't get suggestions of the properties the `user` has.

So basically there's a trade-off between the approach this library takes and what libraries that use transducers use. Now I really really hope that the [pipeline operator proposal](https://github.com/tc39/proposal-pipeline-operator) does get eventually finished and accepted. Then we could simply write that as

```ts
import { filter, sortByDescending, toArray } from 'fromfrom/transforms';

const result = users
  |> sortByDescending(user => user.age)
  |> filter(user => user.active)
  |> toArray
```

and the typings should be more doable.

Now I am thinking of making a more FP oriented version of this same stuff, but it would be a separate library.

**Comment by brucou** [comment]
Thanks for the long answer Tomi. A few comments:

- you can have a functional API and still have your 'fluent' API by using some tricks around proxy. I don't know the story around typings for that though. You can see an example of that around here : https://github.com/langhuihui/rx4rx/blob/master/stdlib/index.js#L403 and https://github.com/langhuihui/rx4rx/blob/master/stdlib/index.js#L407

- transducers are hard to type precisely because they are so generic (and well also because of typescript choices). The advantage on the other hand is that anybody can write its own transducer independently of any transducer library (provided they follow the defined protocol). That may be what you refer to when you say that 'fluent' API are harder to extend.

Your points are valid. It is nice to have autocompletion and the goodies you get with TypeScript. 
I was just thinking that if you already have a proven pattern used in several languages, libraries which have been extensively used (and debugged), why make a new one with its unavoidable specifics and early bugs? I find it better to coalesce around a limited set of generic principles and techniques and reuse rather than redo. In my humble experience, that saves more time than any typings.
